_code/insitucor-custom-module/InSituCor_Module.R

```diff
@@ -0,0 +1,398 @@
+  typeClass <- switch(varType, 
+                      character={"STRING"},
+                      logical={"BOOL"},
+                      numeric={"NUM"})
+  if(required == TRUE){
+    if(!exists(varName)){
+      varMsg <- paste0("\n\"", varName, "\" ", "was not set in custom module creation")
+    }
+  }
+  if(exists(varName)){
+    if(varType == "logical"){
+      if(is.null(get(varName))){
+        assign(varName, FALSE, envir = .GlobalEnv)
+      }
+    }
+    if(!is(get(varName), varType)){
+      if(!is.null(get(varName)) | required == TRUE){
+        varMsg <- paste0(varMsg, paste0("\n\"", varName, "\" varType was not set as ", typeClass, 
+                                        " in custom module creation"))
+      }
+    }
+    if(required == TRUE){
+      if(is.null(get(varName))){
+        varMsg <- paste0(varMsg, paste0("\n\"", varName, "\" was not set and is required"))
+      }else{
+        if(get(varName) == ""){
+          varMsg <- paste0(varMsg, paste0("\n\"", varName, "\" was not set and is required")) 
+        }
+      }
+    }
+  }
+  if(!is.null(varMsg)){
+    varMsg <- paste0(msg, varMsg)
+  }else{
+    varMsg <- msg
+  }
+  return(varMsg)
+}
+
+variableMsg <- NULL
+variableMsg <- variableTest(varName = "celltype", varType = "character", msg = variableMsg)
+
+# stop on variable errors if applicable
+if(!is.null(variableMsg)){
+  stop(variableMsg)
+}
+
+if (verbose) {
+  print("reached2")
+}
+
+#### extract basic data ---------------------------------------
+
+# how to see column names: my_pipeline$tiledbsc_dataset$somas$RNA$obs$attrnames()
+
+counts <- study$somas$RNA$X$members$counts$to_matrix()
+print(str(counts))
+counts <- Matrix::t(counts)
+norm <- sweep(counts, 1, pmax(Matrix::rowSums(counts), 20), "/") * 500
+rm(counts)
+
+if (verbose) {
+  print("reached2.4")
+}
+
+# align norm to metadata:
+annotcellIDs <- study$somas$RNA$obs$to_dataframe("cell_ID")
+annotcellIDs <- annotcellIDs$cell_ID
+
+m <- match(annotcellIDs, rownames(norm))
+norm <- norm[m, ]
+
+# using slide ID as a stand-in for tissue, which won't generally be available
+tissuevec <- study$somas$RNA$obs$to_dataframe("slide_ID_numeric")  
+tissuevec <- tissuevec$slide_ID_numeric
+
+print(table(tissuevec))
+print(head(tissuevec))
+
+# get xy:
+xy <- as.matrix(cbind(study$somas$RNA$obs$to_dataframe("x_slide_mm"),
+                      study$somas$RNA$obs$to_dataframe("y_slide_mm")))
+
+if (verbose) {
+  print("reached3")
+  print(head(xy))
+}
+
+
+
+#### get cell type: ---------------------------------
+
+# (dev note: this doesn't yet allow for the use of refineClusters, or for any
+#  custom cell type assignments. This logic should eventually be expanded to do so.)
+
+celltype <- match.arg(celltype, c("Insitutype", "Leiden"))
+obs_colns <- study$somas$RNA$obs$attrnames()
+if(celltype == "Insitutype"){
+  celltypecolumn <- obs_colns[max(which(grepl("nbclust", obs_colns) & grepl("clusters", obs_colns)))] #(column on the right is the latest)
```

Comment:
just fyi, it might be cleaner & more stringent to use 
`celltypecolumn <- end(grep("^RNA_nbclust_.*_clusters", obs_colns, value = T))[1]`
`celltypecolumn <- end(grep("nn_.*_cluster", obs_colns, value = T))[1]`

Reply:
leaving as-is so I don't have to retest, but noted!